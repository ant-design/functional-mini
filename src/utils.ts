let logSwitch = true;

export const setLogSwitch = (on: boolean) => {
  logSwitch = on;
};

//@ts-expect-error
export const log = (...args) => {
  if (!logSwitch) return;

  console.log.apply(console, args);
};

//@ts-expect-error
export const error = (...args) => {
  console.error.apply(console, args);
};

export const time = (label?: string) => {
  if (!logSwitch) return;

  console.time(label);
};

export const timeEnd = (label?: string) => {
  if (!logSwitch) return;
  console.timeEnd(label);
};

//@ts-expect-error
export const shallowCompare = (obj1, obj2) => {
  if (obj1 === obj2) return true;
  if (typeof obj1 !== typeof obj2) return false;
  if (typeof obj1 !== 'object' || typeof obj2 !== 'object') return false;
  if (obj1 === null || obj2 === null) return false;
  if (Array.isArray(obj1) || Array.isArray(obj2)) return false;
  if (Object.keys(obj1).length !== Object.keys(obj2).length) return false;
  for (const key in obj1) {
    if (obj1[key] !== obj2[key]) return false;
  }
  return true;
};

// 这么做是为了测试方便，隔离一些全局变量
export enum EComponent2Status {
  VALID = 'VALID',
  INVALID = 'INVALID',
  UNKNOWN = 'UNKNOWN',
}

let component2Status = EComponent2Status.UNKNOWN;
export const getComponent2Status = () => {
  return component2Status;
};

export const updateComponent2Status = (status?: EComponent2Status) => {
  if (status) {
    component2Status = status;
    //@ts-expect-error
  } else if (typeof my !== 'undefined') {
    //@ts-expect-error
    component2Status = my.canIUse('component2')
      ? EComponent2Status.VALID
      : EComponent2Status.INVALID;
  }
};

let incrementalId = 0;
const instanceKeyId = '_functional_instance_id_';

//@ts-expect-error
export function getIdFromAppxInstance(appxInstance) {
  if (!appxInstance) throw new Error('appxInstance param is falsy');
  let id;
  // 支付宝端
  if (appxInstance.$id) {
    id = `${appxInstance.$id}-${appxInstance.$page?.$viewId || ''}-${
      appxInstance.is || 'page'
    }`; /* 页面 */
  } else if (appxInstance[instanceKeyId]) {
    id = appxInstance[instanceKeyId];
  } else {
    id = `instance-${incrementalId++}`;
    appxInstance[instanceKeyId] = id;
  }
  if (!id) throw new Error('failed to identify appxInstance.$id');
  return id;
}

export const instanceKeyPropNames = '_functional_instance_prop_names_';

export function mergeObjectKeys(
  ...objects: Record<string, unknown>[]
): string[] {
  let res: string[] = [];
  for (const iterator of objects) {
    res = res.concat(Object.keys(iterator));
  }
  return res;
}
