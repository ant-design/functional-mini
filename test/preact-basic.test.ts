import { expect, test, vi } from 'vitest';
import { delay } from './utils';
import {
  useEffect,
  useState,
  useLayoutEffect,
  act,
  mountElement,
  serverRender,
  createElement,
  VNode,
} from '../src/r';

const h = createElement;

test('act 里的方法都是同步执行完成的', async () => {
  const onceEffect = vi.fn();
  const syncRegisterFn = vi.fn();

  const C = function (p) {
    // 此处 ts 能自动推导出来 s 的类型
    const [s] = useState('init');

    useEffect(() => {
      onceEffect();
    }, []);

    syncRegisterFn();
    return h('div', {}, s);
  };

  act(() => {
    const element = h(C, { foo: 'propsValue' });
    mountElement(element);
  });
  expect(onceEffect).toHaveBeenCalledTimes(1);
  expect(syncRegisterFn).toHaveBeenCalledTimes(1);
});

test('不用 act 就是异步', async () => {
  const onceEffect = vi.fn();
  const syncRegisterFn = vi.fn();

  const C = function (p) {
    const [s] = useState('init');

    useEffect(() => {
      onceEffect();
    }, []);

    syncRegisterFn();
    return h('div', {}, s);
  };

  const element = h(C, { foo: 'propsValue' });
  mountElement(element);

  expect(syncRegisterFn).toHaveBeenCalledTimes(1);
  expect(onceEffect).toHaveBeenCalledTimes(0);
  await delay(10);
  expect(onceEffect).toHaveBeenCalledTimes(1);
});

test('mount 之后触发的事件', async () => {
  const logs: string[] = [];

  const onceEffect = vi.fn().mockImplementation(() => {
    console.log('onceEffect is calling');
  });
  const effectFn = vi.fn();
  const offFn = vi.fn();
  const syncRegisterFn = vi.fn();
  const delayTriggerEffect = 5;
  const C = function (p) {
    const [s, setS] = useState('init');

    useEffect(() => {
      onceEffect();
      logs.push('b');
      setS('d');

      setTimeout(() => {
        logs.push('e');
        setS('effect');
        effectFn.call(undefined);
        logs.push('c');
      }, delayTriggerEffect);

      return offFn;
    }, []);

    syncRegisterFn();

    useLayoutEffect(() => {
      logs.push('d');
    }, []);
    logs.push(`render-${s}`);
    return h('div', {}, s);
  };

  let unmount;
  act(() => {
    const element = h(C, { foo: 'propsValue' });
    unmount = mountElement(element).unmount;
    logs.push('after act');
  });
  // act 里的操作都是同步的 之后同步注册上
  expect(onceEffect).toHaveBeenCalledTimes(1);
  expect(syncRegisterFn).toHaveBeenCalledTimes(2);

  await delay(delayTriggerEffect + 5);
  expect(onceEffect).toHaveBeenCalledTimes(1);
  expect(effectFn).toHaveBeenCalledTimes(1);
  expect(offFn).toHaveBeenCalledTimes(0);

  expect(logs).toEqual([
    'render-init',
    'd',
    'after act',
    'b',
    'render-d',
    'e',
    'c',
    'render-effect',
  ]);

  unmount();
  await delay(30);
  expect(offFn).toHaveBeenCalledTimes(1);
});

test('server render', async () => {
  const C = function (p) {
    useEffect(() => {
      throw new Error('should not fire this effect');
    });
    return h('div', {}, p?.text);
  };
  const instance = h(C, { text: 'value' }) as VNode;
  const data = serverRender(instance);
  expect(data).toMatchInlineSnapshot('"<div>value</div>"');
});
