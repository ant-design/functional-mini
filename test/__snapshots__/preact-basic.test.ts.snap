// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`server render 1`] = `"<div>value</div>"`;
